CHERI TGOT TLS

jrtc27 opened the pull request:
- bsd.mkopt.mk: Allow OPT_FOO to be set to override FOO
- libc/csu: Extract reloc.c include
- rtld-elf: Extract core of tls_get_addr_common into tls_get_block and split
- tls: Expose new _rtld_tls_get_block
- libc: Use new _rtld_tls_get_block interface for dl_iterate_phdr
- rtld-elf: Use new _rtld_tls_get_block interface for dl_iterate_phdr
- rtld-elf: Pass lockstate to allocate_tls
- readelf: Teach about additional NT_CHERI_TLS_ABI variants
- sys: Add new sys/_tls_tgot.h
- sys: Add CHERI ELF gABI TGOT defines
- libc: Add support for CHERI TGOT TLS for static binaries
- rtld-elf: Add CHERI TGOT TLS support
- sys: Support CHERI TGOT TLS in arm64 headers
- libc: Support CHERI TGOT TLS on Morello for static binaries
- rtld-elf: Support CHERI TGOT TLS on Morello
- sys: Support CHERI TGOT TLS in riscv headers
- libc: Support CHERI TGOT TLS on CHERI-RISC-V for static binaries
- rtld-elf: Support CHERI TGOT TLS on CHERI-RISC-V
- cheribsdtest: Expect exact TLS bounds for CHERI TGOT TLS on CHERI-RISC-V
- libc: Support compat CHERI TGOT TLS for static binaries
- libc: Support compat CHERI TGOT TLS on Morello for static binaries
- rtld-elf: Support compat CHERI TGOT TLS
- sys: Support compat CHERI TGOT TLS in arm64 headers
- rtld-elf: Support compat CHERI TGOT TLS on Morello
- bsd.opts.mk: Add new CHERI_TGOT_TLS option

jrtc27 committed:
readelf: Teach about additional NT_CHERI_TLS_ABI variants

jrtc27 committed:
sys: Add new sys/_tls_tgot.h

jrtc27 committed:
libc/csu: Extract reloc.c include

We will need to include this for other orthogonal uses and don't want to
include the implementation twice.

jrtc27 committed:
rtld-elf: Extract core of tls_get_addr_common into tls_get_block and split

The new tls_get_block will be exposed as an API in a future commit,
and both it and tls_get_slot_slow will be used as part of CHERI TGOT TLS
support.

jrtc27 committed:
tls: Expose new _rtld_tls_get_block

This interface lets you query the TLS data block without needing to use
__tls_get_addr with an offset of -TLS_DTV_OFFSET, which won't work once
CHERI TGOT TLS is used, since that offset will become an offset into the
TGOT to load a pointer from, not an offset into the TLS data block.

jrtc27 committed:
libc: Use new _rtld_tls_get_block interface for dl_iterate_phdr

jrtc27 committed:
rtld-elf: Use new _rtld_tls_get_block interface for dlsym

jrtc27 committed:
libc: Add support for CHERI TGOT TLS for static binaries

jrtc27 committed:
sys: Add CHERI ELF gABI TGOT defines

jrtc27 committed:
rtld-elf: Pass lockstate to allocate_tls

This will be needed for CHERI TGOT TLS support, which will need to
perform symbol lookups.

jrtc27 committed:
rtld-elf: Add CHERI TGOT TLS support

jrtc27 committed:
sys: Support CHERI TGOT TLS in arm64 headers

jrtc27 committed:
libc: Support CHERI TGOT TLS on Morello for static binaries

jrtc27 committed:
rtld-elf: Support CHERI TGOT TLS on Morello

jrtc27 committed:
libc: Support CHERI TGOT TLS on CHERI-RISC-V for static binaries

jrtc27 committed:
sys: Support CHERI TGOT TLS in riscv headers

jrtc27 committed:
cheribsdtest: Expect exact TLS bounds for CHERI TGOT TLS on CHERI-RISC-V

jrtc27 committed:
rtld-elf: Support CHERI TGOT TLS on CHERI-RISC-V

jrtc27 committed:
libc: Support compat CHERI TGOT TLS for static binaries

jrtc27 committed:
libc: Support compat CHERI TGOT TLS on Morello for static binaries

jrtc27 committed:
rtld-elf: Support compat CHERI TGOT TLS

jrtc27 committed:
sys: Support compat CHERI TGOT TLS in arm64 headers

jrtc27 committed:
rtld-elf: Support compat CHERI TGOT TLS on Morello

jrtc27 committed:
bsd.opts.mk: Add new CHERI_TGOT_TLS option

jrtc27 committed:
bsd.mkopt.mk: Allow OPT_FOO to be set to override FOO

Just as we allow MK_FOO to be set, overriding any WITH(OUT)_FOO and the
default, we should allow OPT_FOO to be set, overriding any FOO and the
default.

jrtc27 commented:
~> * rtld-elf: Extract part of allocate_tls_offset into allocate_tls_offset_common~
~> * rtld-elf: Track allocated TCBs internally and use for distribute_static_tls~
~> * libc,libthr: Remove __pthread_distribute_static_tls~

~Not part of this PR; based on distribute-static-tls-cleanup.~

> * bsd.mkopt.mk: Allow OPT_FOO to be set to override FOO

Should be upstreamed.

> * libc/csu: Extract reloc.c include

Maybe upstream?

> * rtld-elf: Extract core of tls_get_addr_common into tls_get_block and split
> * tls: Expose new _rtld_tls_get_block
> * libc: Use new _rtld_tls_get_block interface for dl_iterate_phdr
> * rtld-elf: Use new _rtld_tls_get_block interface for dl_iterate_phdr

Should upstream (cleaner interface even ignoring CHERI).

jrtc27 commented on sys/sys/elf_common.h:
```diff
@@ -1103,6 +1109,9 @@ typedef struct {
 #define	R_MORELLO_TLS_TPREL128	59398
 #define	R_MORELLO_FUNC_RELATIVE	59400
 #define	R_AARCH64_FUNC_RELATIVE	59401
+#define	R_MORELLO_TLS_TGOT_SLOT	59916
```
These are currently in the "reserved for private Morello experiments" range; we should reserve real relocations at some point before merging

jrtc27 commented:
I believe this is now ready for review, but one question is whether I should change any of the default ABIs here. For RISC-V I don't see a reason why we couldn't switch over (though will need a .require_clean_build bump), but for Morello I don't think it'd be a good idea to switch (to compat) until we have c18n_policy rebased on top of the latest Morello LLVM as I don't want to be juggling different ABIs between dev and dev_c18n.

bsdjhb commented on share/mk/bsd.mkopt.mk:
```diff
@@ -112,11 +112,13 @@ MK_${var}:=	no
 .if !defined(__${opt}_DEFAULT) || empty(__${opt}_DEFAULT)
 .error __${opt}_DEFAULT undefined or empty
 .endif
+.if !defined(OPT_${var})
```
Is this supposed to be `OPT_${opt}`?

jrtc27 replied:
Yes :( and you're the first one to notice despite https://reviews.freebsd.org/D54876

jrtc27 replied:
I suppose I should take this branch and try and build on Morello with CHERI_TLS=yes to actually test it gets the right value for the host tools...

jrtc27 replied:
Updated

bsdjhb commented on libexec/rtld-elf/rtld.c:
```diff
@@ -4680,9 +4679,8 @@ do_dlsym(void *handle, const char *name, void *retaddr, const Ver_Entry *ve,
-			ti.ti_module = defobj->tlsindex;
-			ti.ti_offset = def->st_value - TLS_DTV_OFFSET;
-			sym = __tls_get_addr(&ti);
+			sym = (char *)_rtld_tls_get_block(defobj->tlsindex) +
```
The log message for this commit should reference dlsym() instead of dl_iterate_phdr()

jrtc27 replied:
Seems so, dl_iterate_phdr was done as part of 09cec6619839b2507342235dd4f2cdebacb4cf96 in the case of rtld

jrtc27 replied:
Updated

bsdjhb commented on lib/libc/gen/tls.c:
```diff
@@ -448,7 +588,24 @@ _init_tls(void)
+			    phdr[i].p_align);
+			libc_tgot_init_size = phdr[i].p_filesz;
+			libc_tgot_init_align = phdr[i].p_align;
+#ifndef __CHERI_PURE_CAPABILITY__
```
Style nit: I have a slight preference for #ifdef vs #ifndef when you have an #else, thus, I'd prefer

```
#ifdef __CHERI_PURE_CAPABILITY__
   // purecap
#else
   // legacy
#endif
```

jrtc27 replied:
Oh this way round was copied from lines 585-590 just above for the equivalent PT_TLS case.

bsdjhb commented on libexec/rtld-elf/map_object.c:
```diff
@@ -537,8 +556,14 @@ obj_free(Obj_Entry *obj)
 {
 	Objlist_Entry *elm;
 
+#ifndef TLS_TGOT
```
A single #ifdef / #else here and elsewhere in this commit?

jrtc27 replied:
This, like many such cases, is because of the compat ABI. Weirdly verbose in isolation, but makes sense when you see the later commits.

bsdjhb commented on libexec/rtld-elf/rtld.c:
```diff
@@ -6355,15 +6819,59 @@ allocate_module_tls(struct tcb *tcb, int index)
 	}
+#endif
 
+#ifndef TLS_TGOT
```
I think the #endif above is also for #ifndef TLS_TGOT so those can just be joined into a single #ifndef block?

bsdjhb commented on libexec/rtld-elf/rtld.h:
```diff
@@ -102,12 +102,24 @@ __BEGIN_DECLS
 extern void *rtld_bind_fptr;
 extern void *tls_get_addr_common_fptr;
 
+#ifndef TLS_TGOT
```
You could minimize the diff here a bit if you chose to interpret tls_last_*, tls_static_* and ld_static_tls_extra as just applying to the TGOT and reusing the same global variable names in both cases.  It's not clear to me that tls_static vs tgot_static in Obj_Entry is really a meaningful distinction either.  The "TLS data" is either static or dynamic whether it's using TGOT or not.  Could also perhaps choose to keep a few function names the same as well (e.g. allocate_tls_offset) though I think the function names matter less in terms of reducing the diff.

jrtc27 replied:
Maybe (I'd be a little worried about upstream changes not being appropriately adapted for the TGOT and silently compiling due to reusing the name, but at the same time TLS rarely gets touched upstream, and most of it these of late has been me), except for the weirdo compat ABI both need to exist for all of these, because static TLS data is in the normal place yet there's also a static TGOT alongside it (at an unknown-at-link-time offset, like IE TLS, even for cases where you could use LE). It looks a bit weird and verbose for the first commit but it means the compat ABI commit can just tweak the #if conditions.

bsdjhb commented on libexec/rtld-elf/aarch64/rtld_start.S:
```diff
@@ -449,5 +450,130 @@ ENTRY(_rtld_tlsdesc_dynamic)
+
+	/* Return (*(dtv->dtv_slots[tlsdesc->tls_index - 1].dtvs_tgot + tlsdesc->tgot_offs)) */
+	ldr	x4, [c2, #24]		/* tlsdesc->tgot_offs */
+	ldr	c3, [c3, x4]
```
Should this be `ldr c0`?  c3 gets overwritten by the next `ldp`?

jrtc27 replied:
Ah, I see what happened. The non-TGOT version has an `scbnds c0, c3, x4` here that I removed but didn't fix the destination register of the original ldr. I wonder how that went unnoticed... I guess not much in base uses TLS outside of libc etc that have -ftls-model=initial-exec, and cheribsdtest's GD TLS tests get optimised by the linker to LE on non-MIPS/RISC-V. I really need to write some more that live in libcheribsdtest_dynamic...

jrtc27 replied:
Updated

bsdjhb commented on libexec/rtld-elf/aarch64/rtld_start.S:
```diff
@@ -449,5 +450,130 @@ ENTRY(_rtld_tlsdesc_dynamic)
+ */
+ENTRY(_rtld_tgot_tlsdesc_static)
+	ldr	x0, [c0, #16]
+	ldr	c0, [c1, x0]
```
I guess this function takes two arguments since this is loading from c1?  Should the comment be adjusted?

jrtc27 replied:
Perhaps, but that's also true of purecap (where it's even weirder, CTPIDR_EL0 gets passed in c2 and preserved, with c1 being a non-argument register that's call-clobbered)...

bsdjhb replied:
I found it confusing at least during review that the comment claimed a single argument.  I would even be happy if we added a note for the purecap case you describe above (but I guess that's for the non-TGOT case on Morello?).

jrtc27 replied:
Now documented (and fixed upstream's mess whilst at it... should upstream that minus the Morello part)

bsdjhb commented on libexec/rtld-elf/aarch64/rtld_start.S:
```diff
@@ -449,5 +450,130 @@ ENTRY(_rtld_tlsdesc_dynamic)
+
+	/* Test fastpath - inlined version of tls_get_addr_common(). */
+	ldr	c2, [c0, #16]
+	ldr	c0, [c1]		/* DTV pointer */
```
Same question about c1 as above.

jrtc27 replied:
In terms of it the calling convention making it an argument (and preserved)?

bsdjhb replied:
Well, and even knowing what it is.  That is, the C function prototype in the comment only lists a single argument, and it would be nice if it listed both of them.

jrtc27 replied:
Now documented as above

jrtc27 committed:
rtld-elf: Reword aarch64 TLSDESC comments to match reality

As well as documenting what Morello does, this also fixes various issues
with upstream's comments.

jrtc27 commented:
buildworld on Morello is now running to check I've got everything right (turns out I had missed the build-tools case in Makefile.libcompat, so lib32 (as the first libcompat) libtinfo's make_hash was built with a TGOT and wouldn't run...)

jrtc27 commented:
Looks good building natively
